ffi/harfbuzz/src/lib.rs

```diff
@@ -69,606 +73,372 @@ use icu_properties::{GeneralCategory, Script};
-    pub fn into_raw(funcs: Self) -> *mut hb_unicode_funcs_t {
-        let ret = funcs.raw;
-        core::mem::forget(funcs);
-        ret
-    }
-
-    /// Borrows the wrapped raw pointer without transferring ownership
-    /// and without affecting the refcount.
-    pub fn as_ptr(&self) -> *mut hb_unicode_funcs_t {
-        self.raw
-    }
-
-    fn empty() -> Result<Self, HarfBuzzError> {
-        let empty = unsafe { hb_unicode_funcs_get_empty() };
-        // The HarfBuzz refcounting convention is that "create"
-        // sets refcount to one, not zero.
-        // https://harfbuzz.github.io/object-model-lifecycle.html
-        let raw = unsafe { hb_unicode_funcs_create(empty) };
-        if raw == empty {
-            Err(HarfBuzzError::Alloc)
-        } else {
-            Ok(Self { raw })
-        }
+impl GeneralCategoryData {
+    /// Construct a new [`GeneralCategoryData`] from a data provider.
+    pub fn try_new_unstable<D>(provider: &D) -> Result<Self, HarfBuzzError>
+    where
+        D: DataProvider<GeneralCategoryV1Marker> + ?Sized,
+    {
+        let gc = maps::load_general_category(provider)?;
+
+        Ok(Self { gc })
     }
 
-    /// Sets up a `hb_unicode_funcs_t` with ICU4X compiled data as the back end as the Unicode
-    /// Database operations that HarfBuzz needs. The `hb_unicode_funcs_t` held
-    /// by the returned `UnicodeFuncs` is marked immutable.
-    ///
-    /// ✨ *Enabled with the `compiled_data` Cargo feature.*
-    ///
-    /// [📚 Help choosing a constructor](icu_provider::constructors)
-    #[cfg(feature = "compiled_data")]
-    pub fn new() -> Result<Self, HarfBuzzError> {
-        let ufuncs = Self::empty()?;
-
-        unsafe {
-            hb_unicode_funcs_set_combining_class_func(
-                ufuncs.raw,
-                Some({
-                    extern "C" fn cb(
-                        _: *mut hb_unicode_funcs_t,
-                        unicode: hb_codepoint_t,
-                        _: *mut c_void,
-                    ) -> hb_unicode_combining_class_t {
-                        CanonicalCombiningClassMap::new().get32(unicode).0
-                            as hb_unicode_combining_class_t
-                    }
-                    cb
-                }),
-                core::ptr::null_mut(),
-                None,
-            );
-        }
+    #[doc = icu_provider::gen_any_buffer_unstable_docs!(ANY, Self::try_new_unstable)]
+    pub fn try_new_with_any_provider(
+        provider: &(impl icu_provider::AnyProvider + ?Sized),
+    ) -> Result<Self, HarfBuzzError> {
+        Self::try_new_unstable(&provider.as_downcasting())
+    }
+    #[cfg(feature = "serde")]
+    #[doc = icu_provider::gen_any_buffer_unstable_docs!(BUFFER,Self::try_new_unstable)]
+    pub fn try_new_with_buffer_provider(
+        provider: &(impl icu_provider::BufferProvider + ?Sized),
+    ) -> Result<Self, HarfBuzzError> {
+        Self::try_new_unstable(&provider.as_deserializing())
+    }
```

Comment:
Nit: Use the helper macro to generate these. Is there a reason you're not?

Reply:
There's no baked constructor so it won't work

Reply:
The macro can be made to not generate baked constructors, we do this when we manually write an infallible baked constructor.

Reply:
I couldn't figure out a way to make it work. You're welcome to try.